### PR TITLE
Example for vtable based upgradeability

### PR DESCRIPTION
**Pull request by spalladino**
This example adds a manually implemented vtable for dispatching calls. Rather than have a single contract implement an implementation version, the registry keeps track of which contract satisfies _each function_ for a version.

This allows to make a minor modification to a function in a contract without having to re-upload the code (see the `should upgrade single function` test). It also drops the limitation of a single contract having to fit within a block, since the actual implementation is spread among multiple contracts.

The main drawback is gas consumption, since now every call requires an additional call to the registry to retrieve the address that implements a particular function. Nevertheless, this could be solved by copying the entire mapping from function signature to implementations to the proxy upon initialization.

### CODE REVIEW

**Comment by facuspagnuolo** [comment]
I like the idea! Would you mind providing a readme file describing the approach? Maybe picking what you already described above is enough :)

**Comment by spalladino** [comment]
@facuspagnuolo done!

File: upgradeability_with_vtable/package.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "zos-lab",
```
Comment (facuspagnuolo) [inline upgradeability_with_vtable/package.json]: `upgradeability-with-vtable` ?

File: upgradeability_with_vtable/package.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  "version": "0.1.0",
+  "private": true,
+  "author": "Francisco Giordano <fran@zeppelin.solutions>",
```
Comment (facuspagnuolo) [inline upgradeability_with_vtable/package.json]: should be you 😛

**Comment by spalladino** [comment]
@facuspagnuolo fixed!

**Comment by AugustoL** [comment]
cool ! great idea !

**Comment by rstormsf** [comment]
This is even better @spalladino 
So, as I understand we can have a registry of contracts and by using `msg.sig` we can differentiate which contract we need to call, correct?

**Comment by spalladino** [comment]
Yes, it's basically splitting the logic among multiple contracts that handle a single shared storage. I imagine this being handled via tooling: you write a single contract, without any size limitations, and then a tool splits it into subcontracts which are deployed and proxied from a vtable.

**Comment by rstormsf** [comment]
@spalladino perfect! I'm going to write my PoC based on this architecture.
I will share my results

**Comment by spalladino** [comment]
@facuspagnuolo should we merge this one...?

**Comment by facuspagnuolo** [comment]
sorry palla I forgot to merge it :)

**Comment by spalladino** [comment]
Jorge Izquierdo pointed us to this similar project: https://github.com/ownage-ltd/ether-router

**Pull request merged**